lib/ecto/query.ex

```diff
@@ -2262,6 +2262,49 @@ defmodule Ecto.Query do
     has_named_binding?(Ecto.Queryable.to_query(queryable), key)
   end
 
+  @doc """
+  Applies the callback function to the query if it does not have a named binding,
+  otherwise returns the query without changes.
+
+  The function accepts an `Ecto.Query` struct and must return the query with an added 
+  named binding.
+
+  ## Examples
+
+      with_named_binding(query, :comments, fn query ->
+        join(query, :left, c in assoc(p, :comments), as: :comments)
+      end)
+
+  For more information on named bindings see "Named bindings" in this module doc.
+  """
+  def with_named_binding(%Ecto.Query{} = query, key, fun) when is_function(fun, 1) do
+    if has_named_binding?(query, key) do
+      query
+    else
+      query
+      |> fun.()
+      |> raise_on_invalid_callback_return(key)
```

Comment:
I made it defensive to protect people from shooting the foot (eg. debugging for hours if you have a typo in the name of a binding and call the function multiple times).

Reply:
FWIW I like what you did. It gives some more focus to the function. It didn't really feel right to me that the anon function could do whatever it wanted.